Test case for scalling down before scale up finished

dshulyak opened the pull request:
Verifies that current pet (one which was created last by scale up action)
will enter running before scale down will take place

related: #30082

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.kubernetes.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/34454)

<!-- Reviewable:end -->

dshulyak commented:
Includes https://github.com/kubernetes/kubernetes/pull/32868

@bprashanth PTAL

erictune commented:
@janetkuo PTAL

janetkuo commented on test/e2e/petset.go:
```diff
@@ -177,6 +209,175 @@ var _ = framework.KubeDescribe("PetSet [Slow] [Feature:PetSet]", func() {
+				}
+				isReady := api.IsPodReady(event.pod)
+				if event.pod.Status.Phase != api.PodRunning || !isReady {
+					framework.Failf("%v expected to be running != %v and ready != %v", event.pod.Name, event.pod.Status.Phase, isReady)
```
since it's OR, it not necessary be "the running != %v and ready != %v"

janetkuo commented on test/e2e/petset.go:
```diff
@@ -177,6 +209,175 @@ var _ = framework.KubeDescribe("PetSet [Slow] [Feature:PetSet]", func() {
+			pst.update(ps.Namespace, ps.Name, func(ps *apps.PetSet) { ps.Spec.Replicas = 1 })
+			pst.restoreProbe(ps, testProbe)
+			for i := len(podTracker) - 1; i >= 0; i-- {
+				event := podTracker[i]
```
Could you explain more on what this loop is doing? Is it looking at the last (and only the last) event on `pet-2` to see if the pod is ready? Do we need to wait before the check?

dshulyak replied:
i wanted to verify that lastPet entered running before scale down started. Will rewrite with watcher

dshulyak replied:
and it is in opposite direction because i wanted to find last event for this pod

bprashanth commented:
@janetkuo thanks! let me know if you need help, otherwise I'll assume you got this

janetkuo commented:
Ping @dshulyak for addressing comments

dshulyak committed:
Validate pet set scale up/down order

This change implements test cases to validate that:

    scale up will be done in order
    scale down in reverse order
    if any of the pets will be unhealthy, scale up/down will halt

Change-Id: I4487a7c9823d94a2995bbb06accdfd8f3001354c

dshulyak committed:
Test case for scalling down before scale up finished

Verifies that current pet (one which was created last by scale up action)
will enter running before scale down will take place

Change-Id: Ib47644c22b3b097bc466f68c5cac46c78dd44552

dshulyak commented:
@janetkuo fixed comments and rebased, please take a look when you will have time

janetkuo commented on test/e2e/petset.go:
```diff
@@ -212,6 +212,64 @@ var _ = framework.KubeDescribe("StatefulSet [Slow] [Feature:PetSet]", func() {
+			pst.breakProbe(ps, testProbe)
+			pst.waitForRunningAndNotReady(2, ps)
+
+			By("scaling down to 1 replica and restoring health of 2nd pod")
```
"Continue scale operations after the 2nd pod, and scaling down to 1 replica"

janetkuo commented on test/e2e/petset.go:
```diff
@@ -212,6 +212,64 @@ var _ = framework.KubeDescribe("StatefulSet [Slow] [Feature:PetSet]", func() {
 		})
 
+		It("scaling down before scale up is finished should wait until current pod will be running and ready before it will be removed", func() {
+			By("creating petset " + psName + " in namespace " + ns)
```
`"Creating stateful set " + psName + " in namespace " + ns + ", and pausing scale operations after each pod"`

janetkuo commented on test/e2e/petset.go:
```diff
@@ -212,6 +212,64 @@ var _ = framework.KubeDescribe("StatefulSet [Slow] [Feature:PetSet]", func() {
+			pst := &statefulSetTester{c: c}
+			pst.waitForRunningAndReady(1, ps)
+
+			By("scaling up stateful set " + psName + " to 3 replicas")
```
... and pausing after the 2nd pod

janetkuo commented on test/e2e/petset.go:
```diff
@@ -212,6 +212,64 @@ var _ = framework.KubeDescribe("StatefulSet [Slow] [Feature:PetSet]", func() {
+			pst.setHealthy(ps)
+			pst.updateReplicas(ps, 1)
+
+			By("verifying that 2nd pod will not be removed if it is not healthy")
```
"Verifying that the 2nd pod will not be removed if it's not running and ready"

janetkuo commented on test/e2e/petset.go:
```diff
@@ -212,6 +212,64 @@ var _ = framework.KubeDescribe("StatefulSet [Slow] [Feature:PetSet]", func() {
+			Expect(err).NotTo(HaveOccurred())
+			pst.restoreProbe(ps, testProbe)
+
+			By("verifying that 2nd pod will enter running and ready before it will be removed by scale down")
```
"Verifying the 2nd pod is not removed by scaling down until it becomes running and ready"

janetkuo commented on test/e2e/petset.go:
```diff
@@ -212,6 +212,64 @@ var _ = framework.KubeDescribe("StatefulSet [Slow] [Feature:PetSet]", func() {
+				},
+			))
+			Expect(err).NotTo(HaveOccurred())
+			pst.restoreProbe(ps, testProbe)
```
Move this down to the below `By`

janetkuo commented:
Only nits, LGTM otherwise. This needs to wait for #32868

dshulyak commented:
@k8s-bot kops aws e2e test this

dshulyak commented:
@k8s-bot gci gke e2e test this

dshulyak commented:
@k8s-bot cvm gke e2e test this

dshulyak commented:
@k8s-bot kops aws e2e test this

janetkuo reviewed (approved):
/lgtm

janetkuo commented:
Adding do-not-merge since this needs to wait for #32868

janetkuo commented:
#32868 is merged
